sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/ViewRegistry.java

```diff
@@ -67,21 +72,92 @@ public static ViewRegistryBuilder builder() {
   }
 
   // Matches an instrument selector against an instrument + meter.
-  private static boolean matchesSelector(
+  private boolean matchesSelector(
       InstrumentSelector selector,
       InstrumentDescriptor descriptor,
       InstrumentationScopeInfo meterScope) {
-    return (selector.getInstrumentType() == null
-            || selector.getInstrumentType() == descriptor.getType())
-        && selector.getInstrumentNameFilter().test(descriptor.getName())
-        && matchesMeter(selector, meterScope);
+    if (selector.getInstrumentType() != null
+        && selector.getInstrumentType() != descriptor.getType()) {
+      return false;
+    }
+    if (selector.getInstrumentName() != null
+        && !matchesName(selector.getInstrumentName(), descriptor.getName())) {
+      return false;
+    }
+    return matchesMeter(selector, meterScope);
+  }
+
+  /**
+   * Determine if the {@code instrumentName} matches the {@code namePattern} from {@link
+   * InstrumentSelector#getInstrumentName()}.
+   *
+   * <p>{@code namePattern} may contain the wildcard characters {@code *} and {@code ?} with the
+   * following matching criteria:
+   *
+   * <ul>
+   *   <li>{@code *} matches 0 or more instances of any character
+   *   <li>{@code ?} matches exactly one instance of any character
+   * </ul>
+   */
+  // Visible for testing
+  boolean matchesName(String namePattern, String instrumentName) {
+    if (!namePattern.contains("*") && !namePattern.contains("?")) {
+      return namePattern.equals(instrumentName);
+    }
+
+    String regex = namePatternCache.computeIfAbsent(namePattern, ViewRegistry::namePatternToRegex);
+
+    return Pattern.matches(regex, instrumentName);
+  }
+
+  /**
+   * Transform the {@code namePattern} to a regex by converting {@code *} to {@code .*}, {@code ?}
+   * to {@code .}, and escaping other regex special characters.
+   *
+   * <p>Based off implementation from: https://www.rgagnon.com/javadetails/java-0515.html
+   */
+  private static String namePatternToRegex(String namePattern) {
```

Comment:
This should return `Pattern` (compiled), not `String`.

Reply:
And we can consider to cache `matchesName` by having it return a `Predicate`, instead of only caching the `Pattern` codepath.

Reply:
Also we can consider just not caching :P IMO a cache signals wanting to fully optimize the codepath so it brings up that sort of point, but for this code we may just want to mark it as not worth optimizing since not on hot path, in such a case I'd recommend not having a cache actually

Reply:
I'm happy to add a cache later if we learn its needed.

I think it would only really make a difference in a situations with lots of views utilizing patterns, and with lots of instruments being added throughout runtime.